---
title: "3' UTR Length Statistics for the Spotter Gar (*Lepisosteus oculatus*)"
author: "Maria Vlachonikolou"
date: "`r Sys.Date()`"
output:
    pdf_document: default

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE)
```
 
This script takes a genome annotation .gtf file as input, extracts the lengths of 3'UTRs and plots their distribution. The reason for this analysis is to put the 3'UTRs lengths into perspective and see if some genes appear as "outliers" with very short 3'UTRs. The output of the GeneExt tool with a extension threshold of 2500 bp is examined to determine its effects. 


Load necessary libraries. 
```{r, message=FALSE}
library(rtracklayer)
library(dplyr)
library(ggplot2)
library(ggridges)
library(viridis)
library(ggridges)
```


Converting .gtf annotation file to a data frame
```{r}
my_obj <- import("~/share/MVlachonikolou/Spotted-Gar/GCF_000242695.1_LepOcu1_genomic.gtf")
class(my_obj)
df<-data.frame(my_obj)
```

Exclusion of unmapped scaffolds and the mitochondrial chromosome (optional). 
```{r}
#df_chrom<-df[grepl("^NC", df$seqnames),]
#df_chrom<-subset(df_chrom, seqnames!="NC_004744.1")
```


Separating by strand. Selecting the transcript and stop codon coordinates and saving them to new dataframes. 
```{r}
# rev
df_rev<-df[grep("\\-", df$strand),]

df_transcript_rev<-df_rev[grepl("transcript", df_rev$type),]

df_stop_rev<-df_rev[grepl("stop_codon", df_rev$type),]

```


```{r}
# for
df_for<-df[grep("\\+", df$strand),]

df_transcript_for<-df_for[grepl("transcript", df_for$type),]

df_stop_for<-df_for[grepl("stop_codon", df_for$type),]

```


Merging the transcript data frame with the stop codon coordinates based on the transcript_id column. Keeping only the relevant information. 
In the .gtf annotation there is a common coordinate system for both forward and reverse strand. For the reverse strand the "start" is the 3' end and the "end" the 5' end, which might be counter-intuitive.  
```{r}
df_trnscrpt_stop_rev<-merge(df_stop_rev, df_transcript_rev[ ,c("transcript_id","start", "end", "width")], by="transcript_id", all=FALSE, suffixes = c("_ORF", "_transcript"))
df_trnscrpt_stop_rev<-df_trnscrpt_stop_rev[ ,c("transcript_id", "seqnames", "end_ORF", "start_transcript", "gene_id", "strand", "width_transcript")]


df_trnscrpt_stop_for<-merge(df_stop_for, df_transcript_for[ ,c("transcript_id", "end", "width")], by="transcript_id", all=FALSE, suffixes = c("_ORF", "_transcript"))
df_trnscrpt_stop_for<-df_trnscrpt_stop_for[ ,c("transcript_id", "seqnames", "end_ORF", "end_transcript", "gene_id", "strand", "width_transcript")]
```


Calculating the 3' UTR length as the distance between the end of the stop codon and the end of the transcript.
Many transcripts are mapped to the same Loci because of alternative splicing. We are selecting here the transcript with the longest ORF and by extend the shortest 3' UTR. 

```{r}
df_trnscrpt_stop_rev$UTR<--1*(df_trnscrpt_stop_rev$start_transcript-df_trnscrpt_stop_rev$end_ORF)
names(df_trnscrpt_stop_rev)[names(df_trnscrpt_stop_rev)=="start_transcript"]<-"end_transcript"


df_trnscrpt_stop_rev_unique <- df_trnscrpt_stop_rev %>%
  group_by(gene_id) %>%
  arrange(desc(end_ORF)) %>%
  filter(row_number() == 1) %>%
  ungroup()

df_trnscrpt_stop_for$UTR<-df_trnscrpt_stop_for$end_transcript-df_trnscrpt_stop_for$end_ORF

df_trnscrpt_stop_for_unique <- df_trnscrpt_stop_for %>%
  group_by(gene_id) %>%
  arrange(desc(end_ORF)) %>%
  filter(row_number() == 1) %>%
  ungroup()

df_trnscrpt_stop<-rbind(df_trnscrpt_stop_for, df_trnscrpt_stop_rev)

```



```{r}
# unique loci only 

df_trnscrpt_stop_unique <- rbind(df_trnscrpt_stop_for_unique, df_trnscrpt_stop_rev_unique)

```

Plotting the data. 
```{r, fig.width=15, fig.height=9}
p3<-ggplot(df_trnscrpt_stop_unique, aes(x=UTR))+
  geom_histogram(bins=50, fill="#F25022", color="#e9ecef", alpha=0.78, boundary = 0) +
    ggtitle("3' UTR Length Distribution - Spotted Gar") +
    geom_vline(aes(xintercept=median(df_trnscrpt_stop_unique$UTR)), colour="#3489eb")+
    xlim(0, 10000)+
    #theme_ipsum() +
    theme(
      plot.title = element_text(size=15)
    )

print(p3)

png("~/share/MVlachonikolou/r_scripts_refined/output_plots/Spotted_Gar_UTR_unique.png", 
  width     = 6.25,
  height    = 3.25,
  units     = "in",
  res       = 1000,
  pointsize = 4)
print(p3)
dev.off()
```


```{r}
cat("The median length of 3'UTRs in the genome annotation of the spotted gar is", median(df_trnscrpt_stop_unique$UTR), "bases .\n")
```


Does the 3'UTR length correlate with the transcript length? 
```{r}
df_trnscrpt_stop_unique$w_i_ratio<-df_trnscrpt_stop_unique$UTR/df_trnscrpt_stop_unique$width_transcript
```


```{r}
p5<- ggplot(df_trnscrpt_stop_unique, aes(x=UTR, y=width_transcript)) + 
    geom_point()+
  xlim(0, 50000)+
    geom_smooth(method=lm , color="red", fill="#69b3a2", se=TRUE) 

print(p5)

png("~/share/MVlachonikolou/r_scripts_refined/output_plots/Spotted_Gar_widthvsUTR.png", 
  width     = 8.25,
  height    = 5,
  units     = "in",
  res       = 1200,
  pointsize = 4)
print(p5)
dev.off()
```







Statistics. 
```{r}
meanutr<-mean(df_trnscrpt_stop_unique$UTR)
medianutr<-median(df_trnscrpt_stop_unique$UTR)



print(meanutr)
print(medianutr)



max(df_trnscrpt_stop_unique$UTR)

sum(df_trnscrpt_stop_unique$UTR==0)

sum(df_trnscrpt_stop_unique$UTR==0)/sum(df_trnscrpt_stop_unique$UTR)


```


```{r}
counts_table <- table(df_trnscrpt_stop_unique$UTR)

all_modes <- as.numeric(names(counts_table[counts_table == max(counts_table)]))

print(all_modes)
```

697+678 end at the stop codon, most probably because there is no information about their 3' UTR available. 



Looking into the GeneExt gtf output file. Repeat of the same process as with the NCBI raw .gtf file. 
```{r}
#setwd("/Users/achinos/Downloads/")
my_obj_2500 <- import("~/share/MVlachonikolou/Spotted-Gar/GeneExtout/Gar547_geneext_result_mv2500_noorph.gtf")
class(my_obj_2500)

df_2500<-data.frame(my_obj_2500)

#table(df$seqnames)
```

```{r}
#df_chrom_2500<-df_2500[grepl("^NC", df_2500$seqnames),]
#df_chrom_2500<-subset(df_chrom_2500, seqnames!="NC_001626.1")
```


The GeneExt .gtf output file does not contain anotations for the stop codon and only has unique transcripts. 
```{r}
# rev
df_rev_2500<-df_2500[grep("\\-", df_2500$strand),]

df_transcript_rev_2500<-df_rev_2500[grepl("transcript", df_rev_2500$type),]

#df_stop_rev_2500<-df_rev_2500[grepl("stop_codon", df_rev_2500$type),]

```

```{r}
df_unique_rev_ext <- merge(df_trnscrpt_stop_rev_unique, df_transcript_rev_2500[, c("gene_id", "start", "three_prime_ext")], by = "gene_id", all = FALSE)


df_unique_rev_ext$UTR_ext<--1*(df_unique_rev_ext$start-df_unique_rev_ext$end_ORF)
names(df_unique_rev_ext)[names(df_unique_rev_ext)=="start"]<-"end_ext"


```


```{r}
# for
df_for_2500<-df_2500[grep("\\+", df_2500$strand),]

df_transcript_for_2500<-df_for_2500[grepl("transcript", df_for_2500$type),]

#df_stop_for_2500<-df_for_2500[grepl("stop_codon", df_for_2500$type),]

```

```{r}
df_unique_for_ext <- merge(df_trnscrpt_stop_for_unique, df_transcript_for_2500[, c("gene_id", "end", "three_prime_ext")], by = "gene_id", all = FALSE)


df_unique_for_ext$UTR_ext<-df_unique_for_ext$end-df_unique_for_ext$end_ORF
names(df_unique_for_ext)[names(df_unique_for_ext)=="end"]<-"end_ext"


```

```{r}
df_ext_utr<-rbind(df_unique_rev_ext, df_unique_for_ext)
df_ext_utr$three_prime_ext <- as.integer(df_ext_utr$three_prime_ext)

```

```{r, fig.width=15, fig.height=9}
p4<-ggplot(df_ext_utr, aes(x=UTR_ext))+
  geom_histogram(bins=90, fill="#69b3a2", color="#e9ecef", alpha=0.98, boundary = 0) +
    ggtitle("-m 2500 3' UTR Length Distribution - Spotted Gar") +
    #geom_vline(aes(xintercept=2500), colour="black")+
    xlim(0, 10000)+
    #theme_ipsum() +
    theme(
      plot.title = element_text(size=15)
    )

print(p4)

png("~/share/MVlachonikolou/r_scripts_refined/output_plots/Spotted_Gar_UTR_2500.png", 
  width     = 10.25,
  height    = 6.25,
  units     = "in",
  res       = 1200,
  pointsize = 4)
print(p4)
dev.off()

```





Looking into the GeneExt gtf output file. Repeat of the same process as with the NCBI raw .gtf file. 
```{r}
#setwd("/Users/achinos/Downloads/")
my_obj_5000 <- import("~/share/MVlachonikolou/Spotted-Gar/GeneExtout/Gar547_geneext_result_mv20000_noorph.gtf")
class(my_obj_5000)

df_5000<-data.frame(my_obj_5000)

#table(df$seqnames)
```

```{r}
#df_chrom_5000<-df_5000[grepl("^NC", df_5000$seqnames),]
#df_chrom_5000<-subset(df_chrom_5000, seqnames!="NC_001626.1")
```


The GeneExt .gtf output file does not contain anotations for the stop codon and only has unique transcripts. 
```{r}
# rev
df_rev_5000<-df_5000[grep("\\-", df_5000$strand),]

df_transcript_rev_5000<-df_rev_5000[grepl("transcript", df_rev_5000$type),]

#df_stop_rev_5000<-df_rev_5000[grepl("stop_codon", df_rev_5000$type),]

```

```{r}
df_unique_rev_ext_5000 <- merge(df_trnscrpt_stop_rev_unique, df_transcript_rev_5000[, c("gene_id", "start", "three_prime_ext")], by = "gene_id", all = FALSE)


df_unique_rev_ext_5000$UTR_ext<--1*(df_unique_rev_ext_5000$start-df_unique_rev_ext_5000$end_ORF)
names(df_unique_rev_ext_5000)[names(df_unique_rev_ext_5000)=="start"]<-"end_ext"


```


```{r}
# for
df_for_5000<-df_5000[grep("\\+", df_5000$strand),]

df_transcript_for_5000<-df_for_5000[grepl("transcript", df_for_5000$type),]

#df_stop_for_5000<-df_for_5000[grepl("stop_codon", df_for_5000$type),]

```

```{r}
df_unique_for_ext_5000 <- merge(df_trnscrpt_stop_for_unique, df_transcript_for_5000[, c("gene_id", "end", "three_prime_ext")], by = "gene_id", all = FALSE)


df_unique_for_ext_5000$UTR_ext<-df_unique_for_ext_5000$end-df_unique_for_ext_5000$end_ORF
names(df_unique_for_ext_5000)[names(df_unique_for_ext_5000)=="end"]<-"end_ext"


```

```{r}
df_ext_utr_5000<-rbind(df_unique_rev_ext_5000, df_unique_for_ext_5000)
df_ext_utr_5000$three_prime_ext_5000 <- as.integer(df_ext_utr_5000$three_prime_ext)

```

```{r, fig.width=15, fig.height=9}
p4<-ggplot(df_ext_utr_5000, aes(x=UTR_ext))+
  geom_histogram(bins=90, fill="#69b3a2", color="#e9ecef", alpha=0.98, boundary = 0) +
    ggtitle("-m 5000 3' UTR Length Distribution - Spotted Gar") +
    #geom_vline(aes(xintercept=5000), colour="black")+
    xlim(0, 30000)+
    #theme_ipsum() +
    theme(
      plot.title = element_text(size=15)
    )

print(p4)

png("~/share/MVlachonikolou/r_scripts_refined/output_plots/Spotted_Gar_UTR_5000.png", 
  width     = 10.25,
  height    = 6.25,
  units     = "in",
  res       = 1200,
  pointsize = 4)
print(p4)
dev.off()

```

```{r, fig.width=10, fig.height=5}
p5<-ggplot(df_ext_utr_5000, aes(x=three_prime_ext))+
  geom_histogram(bins=70, fill="#69b3a2", color="#e9ecef", alpha=0.98, boundary = 0) +
    ggtitle("3' UTR -m 5000 Extension Length Distribution - Spotted Gar") +
    #geom_vline(aes(xintercept=5000), colour="black")+
    xlim(1, 5100)+
    #theme_ipsum() +
    theme(
      plot.title = element_text(size=15)
    )

print(p5)

```



```{r}
#Merge everything to a df 
df_trnscrpt_stop_unique$extension<-"Original"
df_ext_utr$extension<-"2500"
df_ext_utr_5000$extension<- "20000"

colnames(df_trnscrpt_stop_unique)[colnames(df_trnscrpt_stop_unique) == "UTR"] <- "UTR_ext"


all_utr<-rbind(df_trnscrpt_stop_unique[ , c("extension", "UTR_ext", "gene_id")], df_ext_utr[ , c("extension", "UTR_ext", "gene_id")], df_ext_utr_5000[ , c("extension", "UTR_ext", "gene_id")])

all_utr$extension<-factor(all_utr$extension, levels=c("Original", "2500", "20000"))

```



```{r, fig.height=5, fig.width=8}
p0<-ggplot(data=all_utr, aes(x=UTR_ext, group=extension, fill=extension)) +
    geom_density(adjust=1.5, alpha=0.3) +
    xlim(0, 25000)+
  xlab("3' UTR Length")+
  theme(legend.position="top")
#    theme_ipsum()

print(p0)

png("~/share/MVlachonikolou/r_scripts_refined/output_plots/Extension_UTR_comparison.png", 
  width     = 6,
  height    = 3.5,
  units     = "in",
  res       = 1000,
  pointsize = 6)
print(p0)
dev.off()

```


```{r}
ggplot(data=all_utr, aes(x=UTR, group=extension, fill=extension)) +
    geom_density(adjust=1.5) +
    theme_ipsum() +
    facet_wrap(~extension) +
  xlim(0, 25000)+
    theme(
      legend.position="none",
      panel.spacing = unit(0.1, "lines"),
      axis.ticks.x=element_blank()
    )
```














```{r}
data <- data.frame(
  type = c( rep("UTR", nrow(df_ext_utr)), rep("UTR_ext", nrow(df_ext_utr))),
  value = c( df_ext_utr$UTR, df_ext_utr$UTR_ext) 
)
```


```{r, fig.width=12, fig.height=7}
p6<- data %>%
  ggplot(aes(x=value, fill=type)) +
    geom_histogram(bins=90, color="#e9ecef", alpha=0.55, position = 'identity', boundary = 0) +
    ggtitle("3' UTR Original vs -m 5000 Length Distribution - Spotted Gar") +
    scale_fill_manual(values=c("#69b3a2", "#404080")) +
    xlim(0, 10000)+
    #theme_ipsum() +
    #labs(fill="")
    theme(
          plot.title = element_text(size=14)
        )

print(p6)


png("~/share/MVlachonikolou/r_scripts_refined/output_plots/Gar_5000_comparison.png", 
  width     = 10.25,
  height    = 6.25,
  units     = "in",
  res       = 1200,
  pointsize = 4)
print(p6)
dev.off()
```


```{r}
#remove the NAs and 0 
numeric_values <- df_ext_utr$three_prime_ext[!is.na(df_ext_utr$three_prime_ext) & df_ext_utr$three_prime_ext > 0] 

length(numeric_values)
mean(numeric_values)
median(numeric_values)

```





```{r}
sessionInfo()
```

